content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+    {
+        $this->someString = $value;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+```
+
+Let's concentrate, can we modify this object after it has been created? …. Yes we can!
+```php
+$one = new SomeImmutableObject('Pippo');
+echo $one->getValue(); //Pippo
+$one->__construct('Pluto'); 
+echo $one->getValue(); //Pluto
+```
+
+try it! **[break the immutable object] (https://3v4l.org/CiHPG)**
+
+This problem is easy to solve, we must put a flag in the constructor and if it is true throw an exception. 
+_Another way to fix, is to create a `named constuctor` and make the `__constructor` private_. 
+let's do it with the first solution!
+
+# The concrete example, break the immutable object first fix
+
+```php
+<?php declare(strict_types=1);
+
+class SomeImmutableObject
+{
+    private $someString;
+    private $flagCreate = false;
+
+    public function __construct(string $value)
+    {
+        if (true === $this->flagCreate) {
+            throw new \BadMethodCallException('This is an immutable object has already create.');
+        }
+
+        $this->someString = $value;
+        $this->flagCreate = true;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+
+$two = new SomeImmutableObject('Pippo');
+echo $two->getValue(); //Pippo
+$two->__construct('Pluto'); // \BadMethodCallException 
+echo $two->getValue();
+```
+
+Now we are happy and we have our immutable object, are we sure?, mmm no… look here! 
+
+# The concrete example, break the immutable object the second way
+
+```php
+<?php
+class SomeImmutableObject
+{
+    private $someString;
+    private $flagCreate = false;
+
+    public function __construct(string $value)
+    {
+
+        if ($this->flagCreate === true) {
+            throw new \BadMethodCallException('This is an immutable object has already create.');
+        }
+
+        $this->someString = $value;
+        $this->flagCreate = true;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+
+class AnotherClassToBreakImmutableObject extends SomeImmutableObject
+{
+
+    public function __construct(string $value)
+    {
+        $this->someString = $value;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+
+    public function change(): void
+    {
+        $this->someString .= ' and Minny';
+    }
+}
+
+$three = new AnotherClassToBreakImmutableObject('Pippo');
+echo $three->getValue(); //Pippo
+echo $three->change();
+echo $three->getValue(); //the value is: Pippo and Minny -
+```
+try it! **[break the immutable object] (https://3v4l.org/KpnfR)**
+
+It still does not work, ok now we definitely correct promised!
+The problem is that inheritance breaks the encapsulation so a good solution in this case is to use **final**!.
```

Comment:
The problem is that inheritance breaks encapsulation. Therefore, in this case using **final** is a good solution!